### PR TITLE
Sed-1813: implement maven plugin for pipeline integration

### PR DESCRIPTION
**Pull request by iegorov777**

### CODE REVIEW

File: step-functions-plugins/step-functions-plugins-java/step-functions-plugins-java-handler-test/pom.xml [hunk 1]
```diff
@@ -1,62 +1,62 @@
...
-						<goals>
-							<goal>single</goal>
-						</goals>
-					</execution>
-				</executions>
-			</plugin>
-		</plugins>	
-	</build>
-</project>
+<!--
```
Comment (jeromecomte) [inline step-functions-plugins/step-functions-plugins-java/step-functions-plugins-java-handler-test/pom.xml:1]: No apparent change in this file. Ensure that the EOL aren't changed

File: step-junit/pom.xml [hunk 2]
```diff
@@ -1,52 +1,52 @@
...
-						<goals>
-							<goal>test-jar</goal>
-						</goals>
-					</execution>
-				</executions>
-			</plugin>
-		</plugins>
-	</build>
-</project>
+<!-- Copyright (C) 2020, exense GmbH This file is part of STEP STEP is free
```
Comment (jeromecomte) [inline step-junit/pom.xml:1]: No apparent change in this file.

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunDeployedExecutionBundleMojo.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
+package step.plugins.maven;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.Parameter;
+import step.core.repositories.RepositoryObjectReference;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class AbstractRunDeployedExecutionBundleMojo extends AbstractRunExecutionBundleMojo {
```
Comment (jeromecomte) [inline step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunDeployedExecutionBundleMojo.java:10]: ExecutionBundles have been renamed AutomationPackages. Please rename all the classes accordingly

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunDeployedExecutionBundleMojo.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	@Override
+	protected RepositoryObjectReference prepareExecutionRepositoryObject(Map<String, Object> executionContext) {
+		return new RepositoryObjectReference("Artifact", prepareRepositoryParameters());
+	}
+
+	private HashMap<String, String> prepareRepositoryParameters() {
+		HashMap<String, String> repoParams = new HashMap<>();
+		repoParams.put("groupId", getGroupId());
```
Comment (jeromecomte) [inline step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunDeployedExecutionBundleMojo.java:32]: These constants are declared in ArtifactRepository (which we don't want to import here). If we have a suitable common project we could move the constants there

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+			FunctionPackage previousPackage = null;
+			Map<String, String> packageAttributes;
+
+			// init with default if the key is not explicitly defined via maven parameter
+			if (getCustomPackageAttributes() == null || getCustomPackageAttributes().isEmpty()) {
+				packageAttributes = new HashMap<>();
+				packageAttributes.put("artifactId", getProject().getArtifactId());
+			} else {
+				packageAttributes = getCustomPackageAttributes();
```
Comment (jeromecomte) [inline step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java:41]: @david-stephan @iegorov777 We might want to use custom package attributes without using them as tracking attributes. Shouldn't we always use the concatenation of groupId + artifactId as tracking attribute?

File: step-maven-plugin/step-ee-maven-plugin/src/main/java/step/plugins/maven/RunPackagedExecutionBundleMojoEE.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package step.plugins.maven;
+
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import step.client.credentials.ControllerCredentials;
+
+@Mojo(name = "run-packaged-execution-bundle-ee")
+public class RunPackagedExecutionBundleMojoEE extends AbstractRunPackagedExecutionBundleMojo {
```
Comment (jeromecomte) [inline step-maven-plugin/step-ee-maven-plugin/src/main/java/step/plugins/maven/RunPackagedExecutionBundleMojoEE.java:8]: The tenant selection is missing here

**Review by jeromecomte — CHANGES_REQUESTED** [review]
Nice code!
See line comments for required changes

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunExecutionBundleMojo.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+	protected RemoteExecutionManager createRemoteExecutionManager() {
+		return new RemoteExecutionManager(getControllerCredentials());
+	}
+
+	protected abstract RepositoryObjectReference prepareExecutionRepositoryObject(Map<String, Object> executionContext);
+
+	private void checkExecutionRunResult(RemoteExecutionManager remoteExecutionManager, String executionId) throws MojoExecutionException {
+		getLog().info("Waiting for execution result from step...");
+		try {
+			RemoteExecutionFuture executionFuture = remoteExecutionManager.getFuture(executionId).waitForExecutionToTerminate(getExecutionResultTimeoutS() * 1000);
```
[the remark this answers is not in the document]
Comment (iegorov777) [reply step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunExecutionBundleMojo.java:67]: Do you mean we have to add 2 parameters:
- waitForExecution
- ensureExecutionSuccess

And support the following "modes"
- don't wait until execution is finished (waitForExecution=false)
- wait until execution is finished and ensure the status is 'PASSED' (waitForExecution=true, ensureExecutionSuccess=true)
- wait until execution is finished and ignore the final status (waitForExecution=true, ensureExecutionSuccess=false)

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java (hunk 5 above)
Comment (iegorov777) [reply step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java:41]: @jeromecomte probably we can get the situation when want to use versionId additionally to groupId+artifactId
But anyway we can split the configuration parameters for "tracking attributes" and "other package attribute" and apply them independently (if it makes sense).

File: step-maven-plugin/step-ee-maven-plugin/src/main/java/step/plugins/maven/RunPackagedExecutionBundleMojoEE.java (hunk 6 above)
Comment (iegorov777) [reply step-maven-plugin/step-ee-maven-plugin/src/main/java/step/plugins/maven/RunPackagedExecutionBundleMojoEE.java:8]: @jeromecomte do we need to apply the same tenant selection logic as in UploadKeywordsPackageMojoEE, am I right?

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunExecutionBundleMojo.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeromecomte) [reply step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunExecutionBundleMojo.java]: Yes, exactly. The default values should be:
waitForExecution = true
ensureExecutionSuccess = true

File: step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java (hunk 5 above)
Comment (jeromecomte) [reply step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java:41]: @iegorov777 We aligned with @david-stephan. Here the solution we propose:

Per default if getCustomPackageAttributes() is not set, no package attribute has to be set (remove artifactId)
If getCustomPackageAttributes() is set, use it as package attribute (argument the method of newKeywordPackage())

As tracking attribute use a a custom field of the function package: FunctionPackage.customFields.tracking (step.core.accessors.AbstractIdentifiableObject#customFields) and set it to groupId + "." + artifactId. To do so, the method step.functions.packages.client.RemoteFunctionPackageClientImpl#addOrUpdateKeywordPackage might have to be extended

Only this tracking attribute should be used to search for previously uploaded packages

File: step-maven-plugin/step-os-maven-plugin/src/test/java/step/plugins/maven/RunPackagedAutomationPackagesMojoOSTest.java [hunk 8]
```diff
@@ -57,7 +57,7 @@ public void testExecuteOk() throws InterruptedException, TimeoutException, MojoE
 		Mockito.verify(remoteExecutionManagerMock, Mockito.times(1)).execute(captor.capture());
 		ExecutionParameters captured = captor.getValue();
 		Assert.assertEquals("Test description", captured.getDescription());
-		Assert.assertEquals("testUser", captured.getUserID());
+		Assert.assertEquals("dummy", captured.getUserID());
```
[the remark this answers is not in the document]
Comment (david-stephan) [reply step-maven-plugin/step-os-maven-plugin/src/test/java/step/plugins/maven/RunPackagedAutomationPackagesMojoOSTest.java]: was implemented

**Review by david-stephan — APPROVED** [review]

**Review by david-stephan — CHANGES_REQUESTED** [review]

**Review by david-stephan — DISMISSED** [review]

**Review by david-stephan — DISMISSED** [review]

**Review by david-stephan — APPROVED** [review]

**Pull request merged**